### PR TITLE
Update single_source_shortest_path_length and all_pairs_shortest_path_length to return iterators

### PR DESCRIPTION
**Pull request by MridulS**

### CODE REVIEW

File: networkx/algorithms/shortest_paths/generic.py [hunk 1]
```diff
@@ -221,22 +221,24 @@ def shortest_path_length(G, source=None, target=None, weight=None):
...
             else:
                 paths=nx.all_pairs_dijkstra_path_length(G, weight=weight)
         else:
             ## Find paths from all nodes co-accessible to the target.
             with nx.utils.reversed(G):
                 if weight is None:
-                    paths=nx.single_source_shortest_path_length(G, target)
+                    # TODO, for some reason this is not working.
+                    # need to investigate it further.
+                    paths = dict(nx.single_source_shortest_path_length(G, target))
```
Comment (MridulS) [inline networkx/algorithms/shortest_paths/generic.py]: For some reason this is not working. Lot of errors pop up in the connectivity module, everything else is working fine :/

**Comment by MridulS** [comment]
@dschult @hagberg @chebee7i @jfinkels @ysitu

**Comment by MridulS** [comment]
Ok so I found the problem (finally >.<). But I guess my solution is not the best one.

We can't use `nx.utils.reversed` as `nx.single_source_shortest_path_length` returns a iterator. Thus the graph is not reversed when we call `nx.shortest_path_length`.

@dschult @hagberg opinions and suggestions ?

**Comment by hagberg** [comment]
I'm not sure I understand what the problem is with this PR but those changes to digraph.py and multidigraphy.py don't look like the right approach to fixing it.

**Comment by dschult** [comment]
The following is my suggestion, but it may seem too hack-ish for some.

The single_source_shortest_path function will work with ```G.succ``` input instead of ```G```.  So to reverse it you could get away with simply inputting ```G.pred``` instead of ```G```.  It's a hack that relies on the data structure--for sure.

In some sense, the single_source_shortest_path routines could be re-documented to accept any object with a ```__getitem__``` method that returns the neighbors...   It doesn't matter if it is a ```Graph, DiGraph, dict, etc```. just some object that returns neighbors.

Similarly the djikstra routines could be easily rewritten to accept any object that has a method ```edges(u, data="weight", default=1)```.  This would require minor revision to code [no longer check for directed to know whether to look at ```G.succ``` or ```G.adj```... and switch to the ```G.edges(u, data=weight, default=1)``` data access method], but I think the major change would be the documentation.

So the hack I'm suggesting is to input G.pred into the function you are working on.
The proposal I'm suggesting is to rewrite the documentation to ```single_source_shortest_path``` and rewrite ```_djikstra``` code and documentation.  Both would no longer expect an explicitly declared  NetworkX graph class object.

**Comment by chebee7i** [comment]
It seems my comment made its way to the PR. Whoops, the commit showed up by itself on my feed l, for some reason.

**Comment by chebee7i** [comment]
So I'm not against @dschult's proposal, but note that the problem doesn't exist for `shortest_paths` and that is because it consumes the iterator before leaving the reversed context. 

So a very simple solutuon for `shorest_path_lengths` seems to be 1) consume the generator while in the reversed context and 2) make a new iterator over what was consumed. One view is that this is the price you pay for lazy evaluation.

Aside, I think @MridulS's custom reverse is neat. Definitely light-weight and disposable.

**Comment by MridulS** [comment]
@chebee7i I guess this problem will pop up for `shortest_paths` also when we update `single_source_shortest_path` to return an iterator.

**Comment by MridulS** [comment]
I do like @dschult's proposal but it looks hackish and depends on the data structure (dict in this case) as @dschult pointed out. The current code looks pretty. :)
If we decide to do this then I guess we should make it uniform(implemented algorithms and functions expect a graph like object but they will work just fine with `G.adj`, `G.succ` , etc. also ) across the codebase, for starters at least for the `shortest_paths` module.

**Comment by hagberg** [comment]
I still haven't had time to look at this carefully so maybe I am missing the point. 

This seems related to #1696 - perhaps the underlying issue is that an API addition is needed for digraph predecessors to match G[u]?  G.succ[u] and G.pred[u] are maybe not it?  Do we need a G.adjacency() method enhancement?

**Comment by MridulS** [comment]
@hagberg The basic issue is that when we call `nx.shortest_path_length` it returns an iterator which is originally produced by `nx.single_source_shortest_path_length`. And when we use `nx.utils.reversed()` to reverse the graph and reverse it back, it doesn't work.

A simple analogy is given below by `f()` -> `nx.single_......` ,  `g()` -> `nx.shortest_path_length` and list `a` -> `graph`. `nx.utils.reversed` effectively does the same thing as given below, it first reverses the original graph and then reverse it back. 
[I can be wrong about this] When we call `g(a)` the function is executed line by line, starting by `a.reverse()` then `r = f(a)`, but at this point f(a) wont actually return anything yet as it as generator, it will just be placed in the memory. After this `a.reverse()` is called again, thus reversing the list back to original. In the last line when we return `r` this is the time when `f(a)` is actually executed and hence it takes the current state of `a` i.e. the original one.

``` python
In [58]: def f(a):
   ....:     for i in a:
   ....:         yield i
   ....:         

In [60]: def g(a):
   ....:     a.reverse()
   ....:     r = f(a)
   ....:     a.reverse()
   ....:     return r
   ....: 

In [62]: a
Out[62]: [1, 2, 3, 4]

In [63]: list(f(a))
Out[63]: [1, 2, 3, 4]

In [64]: list(g(a))
Out[64]: [1, 2, 3, 4]

```

**Comment by hagberg** [comment]
And we need `reverse` because we are using a bidirectional search?

**Comment by chebee7i** [comment]
Reverse is needed when a target is specified, but not a source. 

@MridulS, have you correctly understood my comment? If you consume the iterator that is returned immediately, for the reverse case only, then you should have no issue.

**Comment by chebee7i** [comment]
That is why `shortest_path` will not have an issue since it must consume to reverse the order of the results.

**Comment by dschult** [comment]
I agree with @chebee7i's fix in this obscure case when target is specified by source is not. Store all the results from the iterator(exhausting it), re-reverse the graph and then prepare to yield the results. It doesn't provide the memory savings of an iterator, but we can't in this case.

shortest_path will also have to exhaust the iterator before flipping the paths. You might be tempted to set up a generator to flip the paths one at a time, but due to the need to exhaust the iterator you're better off just creating the dict and then yielding from that.

@mriduls approach of creating a new graph with monkey-patched data structure works too. Its like an ultra-shallow-copy reverse. The graph and its reverse copy are linked at their core which could cause surprises.

**Comment by MridulS** [comment]
@chebee7i oops I got it now :)
I did the same thing initially https://github.com/networkx/networkx/pull/1715#discussion_r36414081

I think the initial way, adding a `dict()`, is better than `custom` copy way because copying a graph takes up more memory(double). Looks neat too.
We can add a note regarding this behaviour in the code.

**Comment by MridulS** [comment]
dict() or custom copy ?

**Comment by dschult** [comment]
I would prefer dict over custom copy.  But what about using G.pred instead of G as input to single_source_shortest_path?

**Comment by MridulS** [comment]
I do like the G.pred proposal but I think it will break uniformity of using graph object as input to algorithms. If we follow this we should change at least the `shortest_path` module to work like this.

**Comment by MridulS** [comment]
and it looks hackish too

**Comment by dschult** [comment]
OK... let's store the results of the iterator (exhaust it)..  the "dict" method.

**Comment by MridulS** [comment]
I used `list()` to exhaust the iterator to maintain uniformity.
Otherwise we have to use `paths.keys()` / `path.values()` instead of `n for n, d in .....`/ `d for n, d in ....` in `nx.shortest_path_length(G, target = some_target)`

**Comment by dschult** [comment]
Should this get merged into iter_refactor or into master?

**Comment by MridulS** [comment]
It should be merged into master, I'll change that

**Comment by MridulS** [comment]
I have opened a new PR #1739 because I think we can't change the target branch of a PR.

**Pull request closed without merging**